### PR TITLE
Add support for COEP violation reporting

### PR DESCRIPTION
**Pull request by cdumez**
#### 33bacbfe4c429e62f63587f0ba98013a0db2bfad
<pre>
Add support for COEP violation reporting
<a href="https://bugs.webkit.org/show_bug.cgi?id=244985">https://bugs.webkit.org/show_bug.cgi?id=244985</a>

Reviewed by Brent Fulgham.

Add support for COEP violation reporting:
- <a href="https://html.spec.whatwg.org/multipage/origin.html#queue-a-cross-origin-embedder-policy-inheritance-violation">https://html.spec.whatwg.org/multipage/origin.html#queue-a-cross-origin-embedder-policy-inheritance-violation</a>
- <a href="https://fetch.spec.whatwg.org/#queue-a-cross-origin-embedder-policy-corp-violation-report">https://fetch.spec.whatwg.org/#queue-a-cross-origin-embedder-policy-corp-violation-report</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/reporting-navigation.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/reporting-subresource-corp.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/reporting-to-document-reporting-endpoint.https.window-expected.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/reporting/Report.cpp:
(WebCore::Report::createReportFormDataForViolation):
* Source/WebCore/Modules/reporting/Report.h:
* Source/WebCore/Modules/reporting/ReportBody.cpp:
* Source/WebCore/Modules/reporting/ReportingClient.h:
* Source/WebCore/Modules/reporting/ReportingObserver.cpp:
(WebCore::isVisibleToReportingObservers):
* Source/WebCore/Modules/reporting/ViolationReportType.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::sendReportToEndpoints):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/COEPInheritenceViolationReportBody.cpp: Copied from Source/WebCore/Modules/reporting/Report.cpp.
(WebCore::COEPInheritenceViolationReportBody::create):
(WebCore::COEPInheritenceViolationReportBody::COEPInheritenceViolationReportBody):
(WebCore::COEPInheritenceViolationReportBody::disposition const):
* Source/WebCore/loader/COEPInheritenceViolationReportBody.h: Copied from Source/WebCore/Modules/reporting/Report.h.
(WebCore::COEPInheritenceViolationReportBody::blockedURL const):
(WebCore::COEPInheritenceViolationReportBody::encode const):
(WebCore::COEPInheritenceViolationReportBody::decode):
(isType):
* Source/WebCore/loader/COEPInheritenceViolationReportBody.idl: Copied from Source/WebCore/Modules/reporting/ReportBody.cpp.
* Source/WebCore/loader/CORPViolationReportBody.cpp: Copied from Source/WebCore/Modules/reporting/Report.cpp.
(WebCore::CORPViolationReportBody::create):
(WebCore::CORPViolationReportBody::CORPViolationReportBody):
(WebCore::CORPViolationReportBody::type const):
(WebCore::CORPViolationReportBody::disposition const):
(WebCore::CORPViolationReportBody::destination const):
* Source/WebCore/loader/CORPViolationReportBody.h: Added.
(WebCore::CORPViolationReportBody::blockedURL const):
(WebCore::CORPViolationReportBody::encode const):
(WebCore::CORPViolationReportBody::decode):
(isType):
* Source/WebCore/loader/CORPViolationReportBody.idl: Copied from Source/WebCore/Modules/reporting/ReportBody.cpp.
* Source/WebCore/loader/CrossOriginEmbedderPolicy.cpp:
(WebCore::sendCOEPInheritenceViolation):
(WebCore::sendCOEPCORPViolation):
* Source/WebCore/loader/CrossOriginEmbedderPolicy.h:
* Source/WebCore/loader/CrossOriginOpenerPolicy.cpp:
(WebCore::sendViolationReportWhenNavigatingToCOOPResponse):
(WebCore::sendViolationReportWhenNavigatingAwayFromCOOPResponse):
(WebCore::createViolationReportObject): Deleted.
* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendViolationReport):
* Source/WebCore/loader/PingLoader.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::sendReportToEndpoints):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::contextURLforCORPViolation):
(WebKit::performCORPCheck):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::encode const):
(WebKit::NetworkResourceLoadParameters::decode):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::shouldInterruptLoadForCSPFrameAncestorsOrXFrameOptions):
(WebKit::NetworkResourceLoader::shouldInterruptNavigationForCrossOriginEmbedderPolicy):
(WebKit::NetworkResourceLoader::shouldInterruptWorkerLoadForCrossOriginEmbedderPolicy):
(WebKit::NetworkResourceLoader::frameIdentifierForReport const):
(WebKit::NetworkResourceLoader::notifyReportObservers):
(WebKit::NetworkResourceLoader::sendReportToEndpoints):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::ReportBody&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::ReportBody&gt;&gt;::decode):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::addParametersShared):
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
(WebKit::WebLoaderStrategy::startPingLoad):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::sendReportToEndpoints):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/254466@main">https://commits.webkit.org/254466@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/4123615c702534d1ff3f0618028462c7f84f8a58

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89113 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33674 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19943 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98437 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154745 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93122 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32182 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27741 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81473 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92906 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94760 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25555 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76048 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25488 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68463 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29964 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14466 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29692 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15436 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3130 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33136 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38387 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31833 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34529 "Passed tests") | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/failure-check-sequence.https-expected.txt [hunk 1]
```diff
@@ -1,10 +1,6 @@
-CONSOLE MESSAGE: Refused to load https://localhost:9443/common/blank.html?pipe=header(x-frame-options,%20deny) because it does not appear in the frame-src directive of the Content Security Policy.
-CONSOLE MESSAGE: Refused to load https://localhost:9443/common/blank.html because it does not appear in the frame-src directive of the Content Security Policy.
-CONSOLE MESSAGE: Refused to load https://localhost:9443/common/blank.html because it does not appear in the frame-src directive of the Content Security Policy.
-CONSOLE MESSAGE: Refused to display 'https://localhost:9443/common/blank.html?pipe=header(x-frame-options,%20deny)' in a frame because it set 'X-Frame-Options' to 'deny'.
 
 PASS CSP check precedes X-Frame-Options check
 PASS CSP check precedes COEP check - CSP header first
 PASS CSP check precedes COEP check - COEP header first
-FAIL COEP check precedes X-Frame-Options check assert_equals: expected (string) "coep" but got (object) null
+PASS COEP check precedes X-Frame-Options check
```
Comment (brentfulgham) [inline LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/failure-check-sequence.https-expected.txt:5]: Nice!

File: LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/block-local-documents-inheriting-none.https-expected.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
 
-FAIL Prevent local scheme documents from loading within a COEP: require-corp iframe if they inherit COEP: none assert_equals: expected 3 but got 0
+FAIL Prevent local scheme documents from loading within a COEP: require-corp iframe if they inherit COEP: none assert_equals: expected 3 but got 1
```
Comment (brentfulgham) [inline LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/block-local-documents-inheriting-none.https-expected.txt:2]: Some progress, I guess .... :-)

File: LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/reporting-subresource-corp.https-expected.txt [hunk 3]
```diff
@@ -1,30 +1,26 @@
...
+PASS [shared worker] blocked by CORP: same-origin
+FAIL [shared worker] blocked due to COEP assert_equals: expected 2 but got 0
+FAIL [shared worker] blocked during redirect assert_equals: expected 2 but got 0
+PASS [service worker] same-origin
+PASS [service worker] blocked by CORP: same-origin
+FAIL [service worker] blocked due to COEP assert_equals: expected 2 but got 0
+FAIL [service worker] blocked during redirect assert_equals: expected 2 but got 0
+PASS [between service worker and page] same-origin
+PASS [between service worker and page] blocked by CORP: same-origin
+TIMEOUT [between service worker and page] blocked due to COEP Test timed out
```
Comment (brentfulgham) [inline LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/reporting-subresource-corp.https-expected.txt:23]: I wonder what this is due to? And does leaving this with a TIMEOUT slow tests down?

File: LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/failure-check-sequence.https-expected.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+
+PASS CSP check precedes X-Frame-Options check
+PASS CSP check precedes COEP check - CSP header first
+PASS CSP check precedes COEP check - COEP header first
+FAIL COEP check precedes X-Frame-Options check assert_equals: expected (string) "coep" but got (object) null
```
Comment (brentfulgham) [inline LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/failure-check-sequence.https-expected.txt:5]: Interesting. I wonder what this bug is due to.

File: Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp [hunk 5]
```diff
@@ -673,10 +679,14 @@ bool NetworkResourceLoader::shouldInterruptNavigationForCrossOriginEmbedderPolic
...
         auto responseCOEP = WebCore::obtainCrossOriginEmbedderPolicy(response, nullptr);
+        if (m_parameters.parentCrossOriginEmbedderPolicy.reportOnlyValue == WebCore::CrossOriginEmbedderPolicyValue::RequireCORP && responseCOEP.value != WebCore::CrossOriginEmbedderPolicyValue::RequireCORP)
+            sendCOEPInheritenceViolation(*this, m_parameters.parentFrameURL.isValid() ? m_parameters.parentFrameURL : aboutBlankURL(), m_parameters.parentCrossOriginEmbedderPolicy.reportOnlyReportingEndpoint, COEPDisposition::Reporting, "navigation"_s, m_firstResponseURL);
 
         if (m_parameters.parentCrossOriginEmbedderPolicy.value != WebCore::CrossOriginEmbedderPolicyValue::UnsafeNone && responseCOEP.value != WebCore::CrossOriginEmbedderPolicyValue::RequireCORP) {
             String errorMessage = makeString("Refused to display '", response.url().stringCenterEllipsizedToLength(), "' in a frame because of Cross-Origin-Embedder-Policy.");
             send(Messages::WebPage::AddConsoleMessage { m_parameters.webFrameID,  MessageSource::Security, MessageLevel::Error, errorMessage, coreIdentifier() }, m_parameters.webPageID);
+
```
Comment (brentfulgham) [inline Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp]: Nit: Probably don't need this whitespace (you don't use it in the 'shouldInterruptWorkerLoadForCrossOriginEmbedderPolicy')

File: Source/WebCore/Modules/reporting/Report.cpp [hunk 6]
```diff
@@ -62,4 +64,23 @@ const RefPtr<ReportBody>& Report::body()
     return m_body;
 }
 
+Ref<FormData> Report::createReportFormDataForViolation(const String& type, const URL& url, const String& userAgent, const Function<void(JSON::Object&)>& populateBody)
```
Comment (brentfulgham) [inline Source/WebCore/Modules/reporting/Report.cpp:67]: This is much better than my version.

File: Source/WebCore/page/csp/ContentSecurityPolicy.cpp [hunk 7]
```diff
@@ -888,7 +882,7 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
 
     auto report = CSPViolationReportBody::createReportFormDataForViolation(info, usesReportTo, violatedDirectiveList.isReportOnly(), effectiveViolatedDirective, m_referrer, violatedDirectiveList.header(), blockedURI, httpStatusCode);
```
Comment (brentfulgham) [inline Source/WebCore/page/csp/ContentSecurityPolicy.cpp:883]: This should probably be revised to use your 'populateReport' pattern, but obviously not needed for this change.

**Review by brentfulgham — APPROVED** [review]
Looks great!

File: LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/reporting-subresource-corp.https-expected.txt (hunk 3 above)
Comment (cdumez) [reply LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/reporting-subresource-corp.https-expected.txt:23]: It is a bit unfortunate but we do have useful test coverage in this test so I opted to unskip it. When we run tests, we use several instances of WebKitTestRunner so overall time shouldn't be too seriously impacted if one test times out.

File: LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/block-local-documents-inheriting-none.https-expected.txt (hunk 2 above)
Comment (cdumez) [reply LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/block-local-documents-inheriting-none.https-expected.txt:2]: Yes, I'll investigate remaining failures in a follow-up.

File: LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/failure-check-sequence.https-expected.txt (hunk 4 above)
Comment (cdumez) [reply LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/failure-check-sequence.https-expected.txt:5]: This is WK1, no COEP support.

File: Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp (hunk 5 above)
Comment (cdumez) [reply Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp]: Ok, will drop.

File: LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/block-local-documents-inheriting-none.https-expected.txt (hunk 2 above)
Comment (cdumez) [reply LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/block-local-documents-inheriting-none.https-expected.txt:2]: This check that we're getting 1 report for blob, 1 for data and 1 for about:blank. We're only getting the one for blob and not the others. I suspect this is a generic COEP issue and not a reporting issue but I'll investigate further.

**Pull request merged**